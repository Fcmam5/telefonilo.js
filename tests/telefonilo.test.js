const Telefonilo = require('../src/telefonilo');
const { JSDOM } = require('jsdom');
let jsdom = new JSDOM('<!doctype html><html><body></body></html>');
let { window } = jsdom;

// Devices:
const iPad = "Mozilla/5.0 (iPad; CPU OS 4_3_5 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8L1 Safari/6533.18.5";
const galaxy = "Mozilla/5.0 (Linux; Android 6.0.1; SM-G900V Build/MMB29M) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.98 Mobile Safari/537.36";
const ffxOnUbuntuDesktop = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:61.0) Gecko/20100101 Firefox/61.0";

const testHTML = `<!doctype html><html><body><div class="container"><p>If you're on a mobile phone you should click on<span class="phone-num">0123456789</span></p></div></body></html>`;

// Mock userAgent
function useDevice(dvc) {
  global.navigator.__defineGetter__('userAgent', function () {
    return dvc;
  });
}

// Passing encryption function that can be reused for tests.
const goodEncryptionFct = function (x) {
  return x.split ? x.split().map(function (tmp) { return 'not' + tmp }).join() : 'not' + x
}

const getGlobal = () => global;

describe('Test helper functions', () => {
  describe('isMobile()', () => {
    it('should check if the user is on iPhone', () => {
      useDevice(iPad);
      expect(Telefonilo._api.isMobile()).toBe(true);
    });

    it('should check if the user is on Android', () => {
      useDevice(galaxy);
      expect(Telefonilo._api.isMobile()).toBe(true);
    });

    it('should check if the user is NOT on Mobile', () => {
      useDevice(ffxOnUbuntuDesktop);
      expect(Telefonilo._api.isMobile()).toBe(false);
    });
  });

  describe('isValidEncryptionFunction(_function)', () => {
    it('should return false if it`s not a valid function', () => {
      const isValidEncryptionFunction = Telefonilo._api.isValidEncryptionFunction;

      expect(isValidEncryptionFunction()).toBe(false);
      expect(isValidEncryptionFunction('')).toBe(false);
      expect(isValidEncryptionFunction(goodEncryptionFct)).toBe(true);
    });

    it('should return false if it returns the same parameters', () => {
      const isValidEncryptionFunction = Telefonilo._api.isValidEncryptionFunction;
      const wrongEncryptionFct = function (x) { return x; };
      expect(isValidEncryptionFunction(wrongEncryptionFct)).toBe(false);
    });
  });
});

describe('Test Telefonilo', () => {
  it('should be invoked without parameters', () => {
    const telefonilo = Telefonilo();
    expect(telefonilo).toBeDefined();
  });

  it('should match the snapshot for the test html', () => {
    document.body.innerHTML = testHTML;

    useDevice(iPad);
    Telefonilo('.phone-num', true, goodEncryptionFct);
    expect(document.body).toMatchSnapshot();
  })

  it('should throw an error if the encryption function is invalid', () => {
    const wrongEncryptionFct = function (x) { return x; };
    expect(() => { Telefonilo('.phone-num', true, wrongEncryptionFct) }).toThrowError();
  });

  xit('should attach Telefonilo to the global scope', () => {
    expect(true).toBeFalsy();
  });

  it('should work for non-mobile devices', () => {
    document.body.innerHTML = testHTML;

    useDevice(ffxOnUbuntuDesktop);
    Telefonilo('.phone-num', true, goodEncryptionFct);
    expect(document.body).toMatchSnapshot();
  });
});