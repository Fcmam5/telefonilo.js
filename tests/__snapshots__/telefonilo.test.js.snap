// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Telefonilo should match the snapshot for the test html 1`] = `
<body>
  <div
    class="container"
  >
    <p>
      If you're on a mobile phone you should click on
      <span
        class="phone-num"
      >
        <a
          href="tel://not0123456789"
        >
          0123456789
        </a>
      </span>
    </p>
  </div>
</body>
`;

exports[`Test Telefonilo should work for non-mobile devices 1`] = `
<body>
  <div
    class="container"
  >
    <p>
      If you're on a mobile phone you should click on
      <span
        class="phone-num"
      >
        not0123456789
      </span>
    </p>
  </div>
</body>
`;
